submissions/webdevtoday/oop_exercise/index.js

```diff
@@ -0,0 +1,99 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: _put repo URL here_
+   Web app: _put project's github pages URL here_
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+// Define your objects here
+
+class Inhabitant {
+   constructor(name, gender, saying, species, legs, friends) {
+      this.name = name;
+      this.gender = gender;
+      this.species = species;
+      this.saying = saying;
+      this.friends = friends;
+      this.legs = legs;
+   }
+   getSaying() {
+      return this.saying;
+   }
+   getProperties() {
+      return [this.species, this.name, this.gender, this.legs, this.saying, this.friends.join(', ')];
+   }
+   toString() {  
+      return this.getProperties().join('; ');
+   }
+}
+
+class Human extends Inhabitant {
+   constructor(name, gender, saying, friends) {
+      super(name, gender, saying, 'human', 2, friends);
+      this.hands = 2;
+   }
+
+   getProperties() {
+      const props = super.getProperties();
+      props.splice(3, 0, this.hands);
+      return props;
+   }
+}
+
+class Dog extends Inhabitant {
+   constructor(name, gender, friends) {
+      super(name, gender, 'woof-woof!', 'dog', 4, friends);
+   }
+}
+
+class Cat extends Inhabitant {
+   constructor(name, gender, friends) {
+      super(name, gender, 'meow!', 'cat', 4, friends);
+   }
+}
+
+class CatWoman extends Human {
+   constructor(name, saying, friends) {
+      super(name, 'female', saying, friends);
+      this.species = 'CatWoman';
+   }
+}
+
+const dog = new Dog('Jhonny', 'male', ['Andrew', 'Anastasia']);
+const cat = new Cat('Murka', 'female', ['Anastasia']);
+const woman = new Human('Anastasia', 'female', 'I\'m a woman!', ['Andrew', 'Murka', 'Jhonny']);
+const man = new Human('Andrew', 'male', 'I\'m a man!', ['Anastasia', 'Jhonny', 'Murka']);
+const catwoman = new CatWoman('CatWoman', cat.getSaying(), []);
```

Comment:
As we use a saying from a particular cat, our CatWoman is a breed with a particular cat rather than a species on her own.
Alternatively, `Cat` class could have a static property `defaultSaying` (or static method `getDefaultSaying`) that both `Cat` and `CatWoman` classes could use at construction time by referring to `Cat.defaultSaying` or calling `Cat.getDefaultSaying()`.

In fact, composition in es6 classes syntax is a real challenge and cannot be implemented nicely. This is the major learning on composition in this task.

Reply:
> As we use a saying from a particular cat, our CatWoman is a breed with a particular cat rather than a species on her own.
> Alternatively, `Cat` class could have a static property `defaultSaying` (or static method `getDefaultSaying`) that both `Cat` and `CatWoman` classes could use at construction time by referring to `Cat.defaultSaying` or calling `Cat.getDefaultSaying()`.
> 
> In fact, composition in es6 classes syntax is a real challenge and cannot be implemented nicely. This is the major learning on composition in this task.

I wanted to use static method, but I didn't sure

Reply:
No worries. This requirement was intentionally introduced to challenge the students. You took the challenge and solved it.

Your solution, whether it is a borrowing of a particular cat property or Cat class static property, is good enough, given the restrictions that classic OOP approach poses against us.

In fact, if we would employ composition design pattern to build a cat-woman this would require us to have the entire world, all of its inhabitants built by the means of composition.

You most likely will deal with classes+inheritance on your job. Now you know OOP better and will be able to handle pure object-oriented code.

But whenever you have an opportunity to design the entities from the scratch, consider the composition pattern first.